File: src/query/groups.js
import path from "path";
import csv from "papaparse";
import { isEmpty, chunksToLines } from "../stream.js";
import { unescapeNewline } from "../escape.js";

/**
 * Layer 1 — KeyGroupStream.
 * Reads a sorted CSV tablet and yields key groups: { key, values }.
 * Each group collects all values for a contiguous run of the same key.
 * Pure CSV logic, no csvs domain concepts.
 */
export async function* keyGroups(fs, dir, filename) {
  const filepath = path.join(dir, filename);

  const empty = await isEmpty(fs, filepath);

  if (empty) return;

  const lineStream = chunksToLines(fs.createReadStream(filepath));

  let currentKey = undefined;
  let currentValues = [];

  for await (const line of lineStream) {
    if (line === "" || line === "\n") continue;

    const {
      data: [[fstEscaped, sndEscaped]],
    } = csv.parse(line, { delimiter: "," });

    const key = unescapeNewline(fstEscaped ?? "");
    const value = unescapeNewline(sndEscaped ?? "");

    if (currentKey !== undefined && key !== currentKey) {
      yield { key: currentKey, values: currentValues };
      currentValues = [];
    }

    currentKey = key;
    currentValues.push(value);
  }

  if (currentKey !== undefined) {
    yield { key: currentKey, values: currentValues };
  }
}

File: src/dataset/query/groups.rs
use crate::line::Line;
use crate::Result;
use std::fs::File;
use std::path::Path;

/// A group of values sharing the same key in a sorted CSV tablet.
#[derive(Debug, Clone)]
pub struct KeyGroup {
    pub key: String,
    pub values: Vec<String>,
}

/// Layer 1 — reads a sorted CSV tablet and yields key groups.
/// Each group collects all values for a contiguous run of the same key.
/// Pure CSV logic, no csvs domain concepts.
pub fn key_groups(filepath: &Path) -> Result<Vec<KeyGroup>> {
    if std::fs::metadata(filepath).is_err() {
        return Ok(vec![]);
    }

    let mut rdr = csv::ReaderBuilder::new()
        .has_headers(false)
        .flexible(true)
        .from_reader(File::open(filepath)?);

    let mut groups: Vec<KeyGroup> = Vec::new();
    let mut current_key: Option<String> = None;
    let mut current_values: Vec<String> = Vec::new();

    for result in rdr.records() {
        let record = result?;

        let line_escaped = Line {
            key: record.get(0).unwrap_or("").to_owned(),
            value: record.get(1).unwrap_or("").to_owned(),
        };

        let line = line_escaped.unescape();

        match &current_key {
            Some(k) if *k != line.key => {
                groups.push(KeyGroup {
                    key: k.clone(),
                    values: std::mem::take(&mut current_values),
                });
                current_key = Some(line.key);
                current_values.push(line.value);
            }
            None => {
                current_key = Some(line.key);
                current_values.push(line.value);
            }
            _ => {
                current_values.push(line.value);
            }
        }
    }

    if let Some(key) = current_key {
        groups.push(KeyGroup {
            key,
            values: current_values,
        });
    }

    Ok(groups)
}
